pageserver/benches/bench_layer_map.rs

```diff
@@ -33,7 +36,7 @@ fn build_layer_map(filename_dump: PathBuf) -> LayerMap<LayerDescriptor> {
         min_lsn = min(min_lsn, lsn_range.start);
         max_lsn = max(max_lsn, Lsn(lsn_range.end.0 - 1));
 
-        updates.insert_historic(Arc::new(layer));
+        updates.insert_historic(Arc::new(layer), &ctx).unwrap();
```

Comment:
We can plumb `ctx` everywhere if it's needed. But I wonder what particular change in this PR required the need to plumb `ctx`. Why is it unavoidable?

Reply:
if we really need to add `ctx` to all these places, please extract that to a separate commit, for easier review

Reply:
Or a separate PR, but I think this depends on the "should LayerMap receive layers with full hole information or can it calculate them on demand". I think this would be possible to side-step by requiring that the layers have the metadata before being put into the LayerMap.

Reply:
> I do not understand why "ctx" in LayerMap is red flag. RequestContext is actually just a way to propagate task specific context. From my point of view it should be everywhere (in each our method) or never (use LS when it is actually needed).

First reason: The layer map is just a data structure. It should be possible to separate it into its own lib crate. Why? When I rewrote the layer map I spent 1 week writing layer map code and 2 months dealing with the lack of separation between the layer map and the caller (no tests, no documented assumptions, too many requirements of a minimum viable implementation, etc.) What we're doing here with `ctx` is equivalent to plumbing `ctx` into the `HashMap` implementation because we want to read a delta file inside `DeltaLayer::hash`. It's bloated. It's not modular.

Second reason: The need to plumb `ctx` for this PR reveals that the layer map needed some kind of access to some kind of expensive operation worth tracking. When someone reads this code, their guess would be that it either:
a) it reads the holes from the header of the delta layer
b) it reads the entire delta layer to construct the holes
c) it downloads layers (pls no!)

A paranoid reader will assume it's (c) and will have to read a lot more code just to make sure. A careless reader will assume it's (a) and move on. In reality it's (b), so nobody is right. Reading code should not be an adventure. The code should make it obvious what's going on and what can be improved. The reader's attention should be on the important bits. Given how complicated the layer map already is, I'd try to remove ctx completely so that there would be no doubt, but if we have to hack something I'd at least be very explicit with comments and TODOs explaining how it can be done better.

Overall, I agree that if we have `ctx` we should plumb it in all methods proactively. But there are exceptions:
1. We should plumb it in all pageserver code, not all libraries. The layer map is not a library, but it really needs to be (see my reasoning above)
2. In theory, more context is always better. But passing the nuclear launch codes into a `get_timezone` function will raise red flags. We're now passing error handling logic into a function that should never attempt to do any of those things that need the error handling.

It seems like an avoidable problem given that we actually compute the holes explicitly whenever we create a delta layer and add it to the layer map. Why not keep or pass that information?

https://github.com/neondatabase/neon/blob/56af67b5931569525b80256b4a9549d05e2fbc9a/pageserver/src/tenant/timeline.rs#L2422-L2429 They're already computed, but we forget this (compile time) information.

Reply:
> But passing the nuclear launch codes into a get_timezone function will raise red flags.

:laughing:

FWIW, I agree 100% on your reasoning.

Reply:
This shows that it was the right decision to pass `ctx` to all the functions explicitly, instead of using e.g. a task-local variable. It's annoying that we need to pass `ctx` to all the functions - but it's not *literally* all the functions, as this example shows.

Reply:
Sorry, I am really confused.
- All reviewers complain that propagation of context complicates PR and make review more difficult (frankly speaking I do not completely agree - it is quite easy to recognize that change is caused by adding context parameter and it is not necessary to spend a lot of time trying to understand and verify all this similar changes).
- There was suggestion to separate propagation of context to separate commit (or PR?). I have no idea of how it can be implemented.
- @bojanserafimov  wrote that passing context to layer map is "red flag" but agree that "if we have ctx we should plumb it in all methods proactively". Certainly I am not suggesting to add context to `get_timezone`. Certainly it should affect pageserver code. What is the precise criteria? I am not sure. May be if method is calling some other pageserver methods, then doesn;t matter whether them need context or not, it should propagate context.
- @hlinnaka wrote "I agree 100% on your reasoning" (about red flag) and then "this shows that it was the right decision to pass ctx to all the functions explicitly".  So, if it was right decision, then why so many complaints and concerns? I just have to propagate context to all methods where it is required. I think that if we have to propagate context explicitly, then it is better to do it once for all (or most) pageserver functions rather then do it in lazy way - on demand. But I do not insist on it and didn't try to implement such approach in this PR.

May be it is better to discuss propagation of contexts in separate thread?
From my point of view TLS is still much better approach to associate some context with task.
If you wan to make context read-only: it can be easily enforced by Context interface itself (use public getter functions).
Access to TLS requires minimal overhead: there is register pointing to thread-local area, so it just one more indirection.
If we need to somehow transform context to pass to child, it can be easily done with TLS also.
And the argument that "it is always explicit" is IMHO better to add to pros rather  than cons.The nature of TLS is that it task specific data and so it much better fit for association some context with task than explicit passing some parameter which  relation to task is obscure.

Reply:
> It seems like an avoidable problem given that we actually compute the holes explicitly whenever we create a delta layer and add it to the layer map. Why not keep or pass that information?

Hmmm.
In this peace of code we are just inserting new delta layer in layer map.
Information about holes is not yet calculated. It is calculated only in one place - in delta layer by traversing disk B-Tree. It is done when layer is "loaded". When new delta layer is created by frozing in-memory layer, then it is marked as "unloaded" and so holes will be constructed on first access to this layer. In principle it is possible to calculate holes for in-memory layer and then pass it to the created delta layer. But why? What are the benefits? And drawback is code duplication.

But in any case, it is possible to pass holes information when new delta layer is created. But this information is 
not stored anywhere. So when we access layer first time after pageserver restart, we in any case have to extract from layer content.

So either I do not understand your proposal, either it can not be implemented.

Reply:
> Information about holes is not yet calculated.

Line 2242 that I linked to, says `let holes = new_delta.get_holes(ctx)?; `. Is this not calculating holes?

> It is calculated only in one place - in delta layer by traversing disk B-Tree.

Let's just **not** traverse the layer B-Tree while holding the layer map write lock. That's all I'm asking. If it can't be done now because it's difficult, add a TODO and issue for later.

Let's move on in this PR, I'll take out the context from the layer map later if necessary.

Reply:
> Line 2242 that I linked to, says let holes = new_delta.get_holes(ctx)?; . Is this not calculating holes?

Yes, it does. But it is stored in layer's inner. No need to pass it somewhere. Or may be I do not understand you.

> Let's just not traverse the layer B-Tree while holding the layer map write lock. T

It is not done under layer map lock. As I mentioned above it really done in = new_delta.get_holes(ctx)?
where no layer lock is hold.

> Let's address the "reading 256MB under layer map write lock" problem, an

No such problem. And moreover, traversing B-Tree doesn't mean reading all 256Mb. Size of layer index is about 40Mb

Reply:
> It is not done under layer map lock. 

Then why do we need `ctx` under write lock? (Rethorical question, hopefully makes it clear where the code readability problem is)

Ok it's not an **actual** issue, just a readability issue. If it took 10 messages of back and forth to reach this conclusion, it deserves at least a comment in the code. I don't think it's at all obvious.

Reply:
Seems like: "`insert_historic` would usually read the layer to compute holes, but in case of delta insertion that information would already be cached."

Is that the explanation? Or something else?

Even if that's the explanation, it's very fragile to line reordering refactors, and needs an extra comment..

Reply:
> Ok it's not an actual issue, just a readability issue. If it took 10 messages of back and forth to reach this conclusion, it deserves at least a comment in the code. I don't think it's at all obvious.

Sorry, looks like it is just some misunderstanding: I didn't understand your concerns and you didn't understand my explanations. As far as almost all reviewer complain about a lot of changes related with propagation of RequestContext and first of all concentrated on this issue I do not notice your main concern - that it can happen that we we perform some expensive operation (extracting holes) under layer map write lock. Actually it is not true, but it happens unintentionally - actually I didn't realize this problem. I will add comment explaining it. But I want to  notice that this problem isn't somehow related with propagation of request contexts. It is present event ven if `DeltaLyaer:::load()` method doesn't require request context  and I do not need to propagate it everywhere.